[WIP] fts: Dev-only, primitive prototype of text search via pg_search.

mateuszmandera opened the pull request:
Alternative to Postgres built-in text search or pgroonga.

Very crude, early prototype, but in this state it works for doing searches.
- Not deployable in production, since this can't be installed with simple apt install from repo, it's a bit annoying to implement.
- The dev env provisioning is also primitive and probably janky, for now I just did the bare minimum to get things working for myself without having to run a bunch of manual tasks.
- **No highlighting of the matches yet.**
- See the list of TODOs in the code for a bunch more missing stuff.

*Note*: If code comments are mysteriously talking about "tantivy", that is https://github.com/quickwit-oss/tantivy - the search engine that `pg_search` uses and where the query language comes from.

I think the next step here is probably to figure out the production deployment part so that we can do some benchmarking on chat.zulip.org?

mateuszmandera committed:
fts: Dev-only, primitive prototype of text search via pg_search.

Alternative to Postgres built-in text search or pgroonga.

Very crude, early prototype:
- Not deployable in production, since this can't be installed with
simple apt install from repo, it's a bit annoying to implement.
- The dev env provisioning is also primitive and probably janky, for now
I just did the bare minimum to get things working for myself without
having to run a bunch of manual tasks.
- No highlighting of the matches yet.
- See the list of TODOs in the code for a bunch more missing stuff.

mateuszmandera commented on tools/setup/postgresql-init-dev-db:
```diff
@@ -106,6 +106,9 @@ psql -v ON_ERROR_STOP=1 -e -h localhost "$DBNAME_BASE" "$USERNAME" <<EOF
 CREATE SCHEMA zulip;
 EOF
 
+sudo sed -i "s/#shared_preload_libraries = ''/shared_preload_libraries = 'pg_search'/" /etc/postgresql/14/main/postgresql.conf
+sudo systemctl restart postgresql@14-main
```
After the package is installed, `pg_search` needs to be added to preload libraries (https://docs.paradedb.com/deploy/pg_search#update-postgresql-conf). I'm not sure what's a good way to do this in dev?

timabbott replied:
Probably dropping a line into `/etc/postgresql/{pgversion}/main/conf.d/zulip-dev.conf` is the way to go here.

mateuszmandera commented on zerver/lib/narrow.py:
```diff
@@ -755,11 +756,34 @@ def by_group_pm_with(
+            # from the search.
+            return query.where(false())
+
+        condition = literal_column("zerver_message.id").op("@@@")(safe_text_search_query)
```
I believe this is the right way to do the queries in our context - this operator is documented at https://docs.paradedb.com/api-reference/optimization/overview The `fts_index.search(...)` operation shown in most of the docs is stated here as:
> While the [search function](https://docs.paradedb.com/api-reference/full-text/overview) is useful for simple full text search queries, it is not optimized for
>    Returning hundreds of thousands or millions of documents as quickly as possible
>   Usage in complex SQL queries with other “plain SQL” WHERE or JOIN clauses

In the narrows code we're doing a bunch of "plain SQL" filtering, so I think clearly this `@@@` operator is the way to go based on this doc.

A bit more detailed in their 0.10.0 release from last week where they started advertising this operator https://github.com/paradedb/paradedb/releases/tag/v0.10.0:
> using the @@@ operator in a WHERE clause, as opposed to SELECTing directly from idxname.search(...), can lead to drastically improved query execution times in situations where scores (or the effects of scoring on ordering) are not necessary..

timabbott replied:
Cool, yeah I think this is the right design and you should just move a few sentences of that analysis into a comment above this line for posterity.

philippemnoel replied:
Hi guys. In `v0.11.0`, releasing on Tuesday October 15, we are moving fully to the `@@@` operator. It is up to 100X faster on some queries. I would encourage you all to move to it, as `.search` will eventually be deprecated.

mateuszmandera commented on pg_search/migrations/0001_enable.py:
```diff
@@ -0,0 +1,26 @@
+                    schema_name => 'zulip',
+                    table_name => 'zerver_message',
+                    key_field => 'id',
+                    text_fields => paradedb.field('rendered_content',
+                                                  tokenizer => paradedb.tokenizer('en_stem'))
+                                   || paradedb.field('subject', tokenizer => paradedb.tokenizer('en_stem'))
```
These fields can take an additional `fast => true` parameter here to be indexed as "fast fields" (https://docs.paradedb.com/api-reference/indexing/fast_fields), but I don't have a good understanding of whether this is right in our case or not. I suppose we can benchmark with/without `fast => true` on czo.

> A field that is indexed as fast is stored in a column-oriented fashion. Fast fields are necessary for [aggregations/faceting](https://docs.paradedb.com/api-reference/faceting/overview). They can also improve the query times of [filtering](https://docs.paradedb.com/api-reference/full-text/filtering) and BM25 scoring.

timabbott replied:
Hmm. Our current design doesn't intend to use the scoring particularly, but I could imagine us extending the search UI to offer a "Sort by best match" option, and it's probably a pain to rebuild these indices.

The other thing I'm uncertain about is whether we want to continue squashing the subject and content into the same field, even though one is HTML and the other is plain-text, or if there's a better way to implement that detail.

philippemnoel replied:
Fast fields store data in columnar format, rather than row format. This is ideal for fields which you may need to do a COUNT over, for example an `id` field.

Imagine your use case is an e-commerce catalog. You want to show the top 50 results, but tell the user there are X million products in the catalog. Doing a COUNT(id) would allow you to do this. Counts are very slow on row databases like Postgres, so we offer fast fields to store specific fields in columnar format.

mateuszmandera commented on zerver/lib/narrow.py:
```diff
@@ -755,11 +756,34 @@ def by_group_pm_with(
+        safe_text_search_query = literal(text_search_query)
+
+        # TODO: finding the matches for higlighting is not implemented yet
+        query = query.add_columns(
+            literal_column("ARRAY[]::integer[]").label("content_matches"),
+            literal_column("ARRAY[]::integer[]").label("topic_matches"),
+        )
```
`idxname.snippet(...)` can be used for higlighting (https://docs.paradedb.com/api-reference/full-text/highlighting#basic-usage), but it's a little bit tricky because it seems to escape HTML:

```
zulip=> SELECT * from fts_pg_search_idx.snippet('rendered_content:get', highlight_field => 'rendered_content', max_num_chars => 1300) where id=1009;
  id  |                                                                 snippet                                                                 | score_bm25
------+-----------------------------------------------------------------------------------------------------------------------------------------+------------
 1009 | &lt;p&gt;It&#x27;s nice and it &lt;em&gt;feels more modern, but I think&lt;/em&gt; this will take some time to <b>get</b> used to&lt;/p |  3.9742208
(1 row)

zulip=> select id, rendered_content from zerver_message where id=1009;
  id  |                                            rendered_content
------+---------------------------------------------------------------------------------------------------------
 1009 | <p>It's nice and it <em>feels more modern, but I think</em> this will take some time to get used to</p>
(1 row)
```

so that makes it a bit fiddly to be able to correctly map the highlighted positions in the snippet to the right positions in rendered_content. Probably we can just unescape the HTML again, but needs  testing to see if that behaves as well as i hope

timabbott replied:
This seems like a question we should ask the ParadeDB folks about; I'd worry that offsets might be off if the original message contained stuff like `&nbsp;`. And it might be easy for them to just add a variant of this function that doesn't HTML-escape things.

andersk replied:
What we actually want are the `(position, length)` character offsets of each match (as a PostgreSQL multidimensional array `integer[][]`), so we can do the string processing ourselves. We don’t want the database to do the string processing at all.

mateuszmandera replied:
@andersk is your suggestion that we ask upstream to expose a function/option that'll return just the position offsets, instead of the snippet text with html tags inserted?

I think that'd indeed be best. Their current way of inserting html tags where the highlighting should be is consistent with how Postgres does (https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-HEADLINE) - which is why we do that painful processing in `ts_locs_array`. But getting the positions would obviously be superior. They rely on Tantivy (https://github.com/quickwit-oss/tantivy) to generate these snippets, so what they can do here might be constrained by the API tantivy offers, but we can ask.

timabbott replied:
Yes, I think we should ask upstream for that; we got PGroonga to add something similar when we were integrating that, and I think we're not the only consumers of that API.

philippemnoel replied:
We can expose this, it's been requested before. I've put up an issue for it: https://github.com/paradedb/paradedb/issues/1778

If one of you is feeling adventurous, we'd love a contribution! We may not be able to get to it super quickly otherwise.

mateuszmandera commented on zerver/lib/pg_search.py:
```diff
@@ -0,0 +1,48 @@
+        # where each word is a single word. Zulip's interpretation of this
+        # search is "message contains all of these words".
+        #
+        # The tantivy query reflecting that can be concisely formed as
+        # (+word1 +word2 +word3)
+        # as the + syntax makes each word a required match.
```
https://docs.paradedb.com/api-reference/full-text/term#must-term

mateuszmandera commented on zerver/lib/pg_search.py:
```diff
@@ -0,0 +1,48 @@
+        query_string = " ".join(f"+{word}" for word in words)
+        query_string = f"({query_string})"
+
+    final_query = f"rendered_content:{query_string} OR subject:{query_string}"
```
this doesn't quite yet replicate the semantics of our current search, because if you search for `word1 word2`, our **current** search will grab a message if it has `word1` in its content and `word2` in subject. This query as done here, requires to see both words either in content or subject. The query can probably be fixed without too much trouble, but I was trying to keep things basic here for now.

timabbott commented on pg_search/migrations/0001_enable.py:
```diff
@@ -0,0 +1,26 @@
+from django.conf import settings
+from django.db import migrations
+
```
I guess we probably will need to keep built-in `tsearch` as an optional feature, even if we're able to make this to default, because third-party database providers may not offer a database with `pg_search` linked in.

timabbott commented on pg_search/migrations/0001_enable.py:
```diff
@@ -0,0 +1,26 @@
+        ("zerver", "0001_initial"),
+    ]
+
+    database_setting = settings.DATABASES["default"]
```
This looks unused, but I assume the plan is to fill in the `schema_name` field below?

mateuszmandera replied:
was just a mistake, cargo-culting from `pgroonga/migrations/0001_enable.py`. Fixed now

timabbott commented on zerver/lib/pg_search.py:
```diff
@@ -0,0 +1,48 @@
+    # word1 "phrase with spaces" word2 "another phrase"
+    # For now this only supports:
+    # 1) single phrase search: "phrase with spaces"
+    # 2) words search: word1 word2 word3
```
There might be a prep PR to consider in changing how we handle search operands in `add_narrow_conditions` to not squash them together there, and instead have them be separated for this code path. Not sure if clients send them in a way where that'd be helpful; I think that's a protocol issue that maybe got lost from @evykassirer's post-search-transition TODO list -- offering a UI for asking for `word1 word2 "consecutive phrase"` that actually gets sent through and processed properly by the server.
```
    if search_operands:                                                                                 
        is_search = True                                                                                
        query = query.add_columns(topic_column_sa(), column("rendered_content", Text))                  
        search_term = NarrowParameter(                                                                  
            operator="search",                                                                          
            operand=" ".join(search_operands),                                                          
        )                                                                                               
        query = builder.add_term(query, search_term)                                                    
```

evykassirer replied:
I just started a [thread for search followups on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20followups), thanks for the reminder!

mateuszmandera replied:
I'm actually not sure if I'm unaware of some things that might become `search_operands", but searching for the expression
```
"this is a phrase" word1 word2
```

the `search_operands` the server sees in this codepath is:
```
> /srv/zulip/zerver/lib/narrow.py(1097)add_narrow_conditions()
-> search_term = NarrowParameter(
(Pdb) search_operands
['"this is a phrase" word1 word2']
```

So it's just one element and there's nothing to even `" ".join()`. If that's the case and the backend currently only ever just gets a single, big string, then I think it might be fine to leave it as it is (aside of cleaning up the confusing code that makes it like there's a real list involved)? It seems sensible that it would be the backend's job to parse the search query string?

timabbott commented on zproject/default_settings.py:
```diff
@@ -421,6 +421,10 @@
 
+# Whether the server is using the pg_search (paradedb) extension
+# for full-text search.
+USING_PG_SEARCH = False
```
I wonder if the right naming scheme for this is `USING_PARADEDB` or `USING_PG_SEARCH`; maybe a question for upstream as to how they prefer to be referred to in documentation, which is the most prominent place this might appear.

philippemnoel replied:
I would use USING_PARADEDB. We will rename our extension to paradedb eventually.

philippemnoel commented on scripts/lib/build-pgsearch:
```diff
@@ -0,0 +1,10 @@
+
+curl -fLO --retry 3 "https://github.com/paradedb/paradedb/releases/download/v0.10.0/postgresql-14-pg-search_0.10.0-1PARADEDB-jammy_amd64.deb"
+
+dpkg -i "postgresql-14-pg-search_0.10.0-1PARADEDB-jammy_amd64.deb"
```
You should upgrade to 0.10.3. There's some really serious performance improvements (up to 100x faster queries). We're releasing `v0.11.0` on Tuesday October 15, and it'll have even more improvements

philippemnoel commented on tools/setup/postgresql-init-dev-db:
```diff
@@ -116,6 +119,8 @@ SELECT unnest(
     END
 ) \gexec
+
+CREATE EXTENSION pg_search;
```
Note that we recommend also installing `vector` (pgvector) if you want to do hybrid search: https://docs.paradedb.com/documentation/guides/hybrid

mateuszmandera committed:
paradedb: Only use paradedb if search query starts with !parededb.

This makes prod testing easier by allowing comparison between results
wiht or without paradedb. This commit should be dropped before merging
obviously.

mateuszmandera commented:
pushed upgraded to pg_search 0.13.1, since it has some important bugfixes
